thanos to aggregator migration doc (#865)
kubecost/docs: install-and-configure/install/multi-cluster/federated-etl/thanos-migration-guide.md

# Migration Guide from Thanos to Kubecost 2.0 (Aggregator)

This tutorial is intended to help our users migrate from the legacy Thanos federation architecture to [Kubecost 2.0 (Aggregator)](aggregator.md). There are a few requirements in order to successfully migrate to Kubecost 2.0. This new version of Kubecost includes a new backend Aggregator which handles the ETL data built from source metrics more efficiently. Kubecost 2.0 provides new features, optimizes UI performance, and enhances the user experience.

## Key changes

* Assets and Allocations are now paginated using `offset`/`limit` parameters
* New data available for querying every 2 hours (can be adjusted)
* Substantial query speed improvements even when pagination not in effect
* Data ingested into and queried from Aggregator component instead of directly from bingen files

> * Idle (sharing), Cluster Management sharing, and Network are computed a priori

Reviewer: I think this is an unnecessary technical detail, let me know if you disagree

Reviewer 2: Specifically the line about Idle? I agree! I think its worthwhile to keep the other lines though.

Reviewer: Yes, just the idle line. Agreed on the rest

Reviewer 2: Updated

{% endhint %}

```sh
kubectl create secret generic federated-store --from-file=federated-store.yaml -n kubecost
```

### Step 5: Set Aggregator and ETL Utils in *values.yaml* on the primary cluster

{% hint style="warning" %}
*Do not* disable Thanos during this step. This will not negatively impact the source data. Thanos will be disabled in a later step.
{% endhint %}

* [Enable Aggregator](https://docs.kubecost.com/install-and-configure/install/multi-cluster/federated-etl/aggregator) for the primary cluster. Your ETL Utils configuration should look like:

```yaml
etlUtils:
  enabled: true
  thanosSourceBucketSecret: kubecost-thanos
  fullImageName: gcr.io/kubecost1/cost-model-etl-utils:0.2
  resources: {}
kubecostModel:
  federatedStorageConfigSecret: federated-store
```

### Step 6: Apply the changes and wait for data to populate in the UI

Ensure all pods and containers are running:

```txt
kubecost-cost-analyzer-685fd8f677-k652h        4/4     Running   2 (12m ago)   3h2m
kubecost-forecasting-6cf4dd7b98-7z8f5          1/1     Running   0             66m
kubecost-etl-utils-6cdd489596-5dl75           1/1     Running   0          6d20h
```

> 

Reviewer: There's another thread open for this, but I want to make doubly-sure we address it: this is not the expected state. It's more like this:

```
→ k get pod -n kubecost-cloudcost
NAME                                                    READY   STATUS    RESTARTS     AGE
kubecost-cloudcost-aggregator-0                         1/1     Running   0            16h
kubecost-cloudcost-etl-utils-79b89f66-v5jq2             1/1     Running   0            3d22h
kubecost-cloudcost-cost-analyzer-7f9b79d74-rmrrt        2/2     Running   0            3d16h
```

Reviewer 2: @michaelmdresser Something I'll clarify in this PR is that if you're migrating off Thanos, you need to perform all these steps before even upgrading up to 2.0. This is necessary since Thanos is not included in the 2.0 Helm Chart. Therefore the 4 containers in `kubecost-cost-analyzer` in v1.108 may be the expected state? I don't expect forecasting though.

I'm going to focus on refining & publishing this doc today. I'll update this code block!

Reviewer: Great clarification, that makes excellent sense

# Migration Guide from Thanos to Kubecost 2.0 (Aggregator)

This tutorial is intended to help our users migrate from the legacy Thanos federation architecture to [Kubecost 2.0 (Aggregator)](aggregator.md). There are a few requirements in order to successfully migrate to Kubecost 2.0. This new version of Kubecost includes a new backend Aggregator which handles the ETL data built from source metrics more efficiently. Kubecost 2.0 will provide new features, optimize the Kubecost's UI performance, and enhance the user experience.

## Key changes
* Assets and Allocations are now paginated

>    * Frontend/API users never have full set of information

Reviewer 3: Can someone explain this a little better? FE + API users encompasses all Kubecost users right?

Reviewer 4: so, I guess there is an important distinction here. The FE users don't have the full set of information, because the FE uses pagination. That's why we now have a separate 'topline' call that gives them the totals for the top line display. 

However - if you leave off the pagination parameters (limit= and offset=), then the endpoints default to returning _everything_ . So I guess an API user that did not include limit/offset would get everything back, the full information set

Reviewer 3: Didn't we make limit/offset mandatory for certain APIs? Also, haven't FE users always had pagination? We just recently changed the name of the parameter for it

Reviewer: I would remove this bullet entirely. "Frontend/API users never have full set of information" isn't a correct statement; it is enough to say that there is now pagination and (below) offset/limit govern this.

   * Use `offset`/`limit` parameters to govern pagination
* New data available for querying every 2 hours (can be adjusted)
* Embedded DuckDB database serves queries
   * Data no longer queried directly from bingen files
   * Substantial query speed improvements even when pagination not in effect
* Data ingested into independent Aggregator component
* Idle (sharing), Cluster Management sharing, and Network are computed a priori
* Distributed tracing integrated into core workflows
* No more pre-computed "agg stores"
   * Request-level caching still in effect

<details>

<summary>Diagram of Aggregator:</summary>

![aggregator-diagram](/images/aggregator/aggregator-diagram.png)

</details>

## Migration Path 

<details>

<summary>Diagram of Aggregator:</summary>

![migration-diagram](/images/aggregator/migration-diagram.png)

</details>

> * All steps are done on the primary except for step 8.

Reviewer 3: Am I misinterpreting this? Step 8 is specifically done on the primary cluster

Reviewer 3: Step 7 is also done on the secondary clusters

Reviewer 2: I've went ahead and removed these two bullet points to reduce any confusion:

> * All steps are done on the primary cluster except for step 8.
> * Nothing needs to be done on the secondary clusters for the migration to be successful. The Thanos sidecar on the secondary clusters will have no impact on the migration or functionality.

* Nothing needs to be done on the secondary clusters for the migration to be successful. The Thanos sidecar on the secondary clusters will have no impact on the migration or functionality.
* Once Aggregator is enabled, all queries hit the Aggregator container and NOT cost-model via the reverse proxy.
* ETL Utils does not destroy the Thanos data, it creates additional directories in the object store.
* For larger environments, the StorageClass must have 1GBPS throughput.
* Having enough storage is important and will vary based on environment.

## Prerequisites

To migrate from Thanos multi-cluster federated architecture to Aggregator, users *must* complete the following steps:

> ### Step 1: Use the existing Thanos object store or create a new dedicated object store

Reviewer 3: Is this even a step? We need to show users how to create a new object store if they prefer that option.

Reviewer: I think the right way to clarify this is that users of metric federation tools other than Thanos which don't require an object store (e.g. AMP) will not have a bucket yet and thus will need to create one.

Reviewer 2: I just addressed this by adding some references to our long-term-storage docs. So users who were on tools like AMP have instructions to set up a bucket.

```
kubecostModel:
  etlBucketConfigSecret: <YOUR_SECRET_NAME>
```

### Step 3: Validate that an `/etl` directory is present in the object store

There should be ETL data present in the following directories. CloudCosts will only have ETL data if cloud integration is enabled.

* `/etl/bingen/allocations`
* `/etl/bingen/assets/`
* `/cloudcosts/`

### Step 4: Create a new secret for the federated store

This will point to the existing thanos object store or the new one created in Step 1.

{% hint style="warning" %}
The name of the .yaml file used to create the secret *must* be named _federated-store.yaml_ or Aggregator will not start.
{% endhint %}

```
kubectl create secret generic federated-store --from-file=federated-store.yaml -n kubecost
```

### Step 5: Set Aggregator and ETL Utils in *values.yaml* on Primary Kubecost Instance.

{% hint style="warning" %}
*Do not* disable Thanos during this step. This will not negatively impact the source data. Thanos will be disabled in a later step.
{% endhint %}

* [Enable Aggregator (primary kubecost instance only)](https://docs.kubecost.com/install-and-configure/install/multi-cluster/federated-etl/aggregator). 

Note:  

* ETL Utils
```

> etlUtils:
>   thanosSourceBucketSecret: kubecost-thanos
>   affinity: {}
>   enabled: true
>   env:
>     LOG_LEVEL: debug
>   fullImageName: gcr.io/kubecost1/cost-model-etl-utils:0.2
>   nodeSelector:
>     kubernetes.io/arch: amd64
>   resources: {}
>   tolerations: {}
> kubecostModel:
>   federatedStorageConfigSecret: kubecost-thanos

Reviewer 2: ```suggestion
etlUtils:
  thanosSourceBucketSecret: kubecost-thanos
  enabled: true
  fullImageName: gcr.io/kubecost1/cost-model-etl-utils:0.2
  resources: {}
kubecostModel:
  federatedStorageConfigSecret: federated-store
```

Federated ETL secret needs to be `federated-store`

kubecost-forecasting-6cf4dd7b98-7z8f5          1/1     Running   0             66m
kubecost-etl-utils-6cdd489596-5dl75           1/1     Running   0          6d20h
```

This can take some time depending on how many unique containers are running in the environment and how much ETL data is in the object store. A couple importants steps are happening in the background which take time to complete.

* The ETL Utils image is building the directory structure in the object store needed by Aggregator to pull the ETL data. 
* SQL tables are building.

Ensure all data loads into the UI before moving onto step 7.

### Step 7: Remove Thanos sidecar from Secondary Clusters.

* Use the same federated-store secret created in step 4 and add the following to the values.yaml for the secondary clusters:

```
federatedETL:
  useExistingS3Config: false
  primaryCluster: false
  federatedCluster: true
kubecostModel:
  containerStatsEnabled: true
  federatedStorageConfigSecret: federated-store
  warmCache: false
  warmSavingsCache: false
```

* Remove the [thanos manifest](https://raw.githubusercontent.com/kubecost/cost-analyzer-helm-chart/v1.108.1/cost-analyzer/values-thanos.yaml)

> ### Step 8: Remove Thanos Configuration from Primary.

Reviewer 3: Depending on how the user is applying changes to their values-thanos.yaml, will they need to perform a `helm` upgrade at the end to validate the changes?

Author: That's one method - the common use case we see with our customers is through CI/CD (e.g. argo, flux, etc). If we want to standardize this per our other examples throughout our documentation we can and users will get the idea since helm integrates with these platforms.

# Migration Guide from Thanos to Kubecost 2.0 (Aggregator)

> This tutorial is intended to help our users migrate from the legacy Thanos federation architecture to [Kubecost 2.0 (Aggregator)](aggregator.md). There are a few requirements in order to successfully migrate to Kubecost 2.0. This new version of Kubecost includes a new backend Aggregator which handles the ETL data built from source metrics more efficiently. Kubecost 2.0 will provide new features, optimize the Kubecost's UI performance, and enhance the user experience.

Reviewer: ```suggestion
This tutorial is intended to help our users migrate from the legacy Thanos federation architecture to [Kubecost 2.0 (Aggregator)](aggregator.md). There are a few requirements in order to successfully migrate to Kubecost 2.0. This new version of Kubecost includes a new backend Aggregator which handles the ETL data built from source metrics more efficiently. Kubecost 2.0 provides new features, optimizes UI performance, and enhances the user experience.
```

## Key changes

* Assets and Allocations are now paginated
   * Frontend/API users never have full set of information
   * Use `offset`/`limit` parameters to govern pagination
* New data available for querying every 2 hours (can be adjusted)
* Embedded DuckDB database serves queries
   * Data no longer queried directly from bingen files

>    * Substantial query speed improvements even when pagination not in effect

Reviewer: This is all correct, but I don't think it should matter to a user that we have an embedded DuckDB database, all they should care about is the last bullet "substantial speed improvements".

Author: @michaelmdresser - thanks for chiming in here. Most current customers I've been discussing 2.0 with have been asking for details about the architectural changes, breaking changes, etc. I don't think this needs to go in our formal 2.0 doc(s) but I think it'd be useful here. Perhaps we don't need to mention its a DuckDB database specifically? Maybe it's not relevant enough?

Reviewer: I'd rather not mention what database -- databases are a touchy topic for many Kubernetes users. The fact that we're using an embedded database mitigates many of those concerns, but I don't want to have to get into the weeds to explain it. I'd go for something more like "overhauled query execution engine which efficiently combines multi-cluster data and serves it to queries faster than ever before". That wording could use some work, but that's the idea we want to convey. Then we can just lean on decent architecture diagrams (more to do there!) to show the high-level changes.

Author: @thomasvn lgtm. Thanks!

Reviewer: LGTM, thanks Thomas

* Data ingested into independent Aggregator component

> * Idle (sharing), Cluster Management sharing, and Network are computed a priori

Reviewer: Another technical detail. IMO this shouldn't matter to a user (its just a technical trick for performance) but if you've gotten questions about this its fine to include.

* Distributed tracing integrated into core workflows

> * No more pre-computed "agg stores"

Reviewer: ```suggestion
* No more pre-computed "AggStores"; this reduces the memory footprint of Kubecost
```

   * Request-level caching still in effect

<details>

<summary>Diagram of Aggregator:</summary>

![aggregator-diagram](/images/aggregator/aggregator-diagram.png)

</details>

## Migration path 

<details>

<summary>Diagram of Aggregator:</summary>

![migration-diagram](/images/aggregator/migration-diagram.png)

</details>

* All steps are done on the primary cluster except for step 8.
* Nothing needs to be done on the secondary clusters for the migration to be successful. The Thanos sidecar on the secondary clusters will have no impact on the migration or functionality.
* Once Aggregator is enabled, all queries hit the Aggregator container and NOT cost-model via the reverse proxy.
* ETL Utils does not destroy the Thanos data, it creates additional directories in the object store.
* For larger environments, the StorageClass must have 1GBPS throughput.

> * Having enough storage is important and will vary based on environment.

Reviewer: ```suggestion
* Having enough storage is critically important and will vary based on environment. Bias towards a larger value for `aggregatorDBStorage.storageRequest` at the beginning and cut it down if persistent low utilization is observed.
```

## Prerequisites

To migrate from Thanos multi-cluster federated architecture to Aggregator, users *must* complete the following steps:

### Step 1: Use the existing Thanos object store or create a new dedicated object store

> This object store is where the ETL backups will be pushed to from the primary cluster.

Reviewer: ```suggestion
This object store is where the ETL backups will be pushed to from the primary cluster's cost-model.
```

### Step 2: Enable [ETL Backups](https://docs.kubecost.com/install-and-configure/install/etl-backup#google-cloud-storage) on the *primary cluster only*. 

> This will ensure we have all historical data in durable storage and it will create the directory the ETL Utils container needs in order to create the directory structure needed for Aggregator to pull the ETL from each respective cluster.

Reviewer: ```suggestion
This ensures we persist historical data in durable storage (outside of Thanos) and stores the data in a format consumable by the ETL Utils container. The ETL Utils container transforms that data and writes it to a separate location in the object store for consumption by Aggregator.
```

```
kubecostModel:
  etlBucketConfigSecret: <YOUR_SECRET_NAME>
```

### Step 3: Validate that an `/etl` directory is present in the object store

> There should be ETL data present in the following directories. CloudCosts will only have ETL data if cloud integration is enabled.

Reviewer: ```suggestion
There should be ETL data present in the following directories. CloudCosts will only have ETL data if at least one cloud integration is enabled.
```

```

### Step 5: Set Aggregator and ETL Utils in *values.yaml* on the primary cluster.

{% hint style="warning" %}
*Do not* disable Thanos during this step. This will not negatively impact the source data. Thanos will be disabled in a later step.
{% endhint %}

* [Enable Aggregator for the primary cluster](https://docs.kubecost.com/install-and-configure/install/multi-cluster/federated-etl/aggregator). 

Note:  

* ETL Utils
```
etlUtils:
  thanosSourceBucketSecret: kubecost-thanos
  affinity: {}
  enabled: true
  env:
    LOG_LEVEL: debug
  fullImageName: gcr.io/kubecost1/cost-model-etl-utils:0.2
  nodeSelector:
    kubernetes.io/arch: amd64
  resources: {}
  tolerations: {}
kubecostModel:
  federatedStorageConfigSecret: kubecost-thanos
```

### Step 6: Apply the changes and wait for data to populate in the UI. 

Ensure all pods and containers are running:

 ```
kubecost-cost-analyzer-685fd8f677-k652h        4/4     Running   2 (12m ago)   3h2m
kubecost-forecasting-6cf4dd7b98-7z8f5          1/1     Running   0             66m
kubecost-etl-utils-6cdd489596-5dl75           1/1     Running   0          6d20h
```

> This can take some time depending on how many unique containers are running in the environment and how much ETL data is in the object store. A couple importants steps are happening in the background which take time to complete.

Reviewer: I need to confirm this, but I'm pretty sure that Aggregator has to be running in a separate Pod in order for it to target "uncombined" data in the object store. When running in the single-pod config (within kubecost-cost-analyzer) I think it only read from the local directory. The end-state we want users to be in here is an "enterprise" setup with separate Aggregator StatefulSet.

Author: @michaelmdresser gotcha - I thought it would be one of the 4 containers in the cost-analyzer pod. Not sure if thats been finalized.

Reviewer: That's the default configuration, but when using an object store for metric federation the config switches to having Aggregator running separately from cost-analyzer, sort of like this:
```
→ k get pod -n kubecost-cloudcost
NAME                                                    READY   STATUS    RESTARTS     AGE
kubecost-cloudcost-aggregator-0                         1/1     Running   0            16h
kubecost-cloudcost-cloud-cost-79b89f66-v5jq2            1/1     Running   0            3d22h
kubecost-cloudcost-cost-analyzer-7f9b79d74-rmrrt        2/2     Running   0            3d16h
kubecost-cloudcost-diagnostics-59f98b8dc9-mv5gz         1/1     Running   0            15h
kubecost-cloudcost-forecasting-6584f99d8d-np46p         1/1     Running   0            15h
```

If that's unclear let's dig more, I want to make sure folks are on the same page about this.

* The ETL Utils image is building the directory structure in the object store needed by Aggregator to pull the ETL data. 

> * SQL tables are building.

Reviewer: What is this supposed to mean?

Author: @michaelmdresser - attempted to come up with some wording around what's happening in the background once aggregator and etl utils is deployed. My understanding and experience is it takes some time for data to come up in the UI. I think adding a banner in the UI letting the user know to wait would be ideal.

Reviewer: Got it, I think we're testing some approaches to such a banner in eng right now -- I'm not sure if that will make 2.0 yet but its definitely a priority.
